controllers/storageautoscaler/scraper.go

```diff
@@ -0,0 +1,149 @@
+package storageautoscaler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+// StorageAutoscalerScrpaer is the scraper for Prometheus metrics
+type StorageAutoscalerScraper struct {
+	client.Client
+	Log               logr.Logger
+	OperatorNamespace string
+	ScrapeInterval    time.Duration
+	SyncMap           *sync.Map
+	EventCh           chan event.GenericEvent
+}
+
+func (r *StorageAutoscalerScraper) ScrapeMetricsPeriodically(ctx context.Context) {
+	for {
+		select {
+		// context of the main reconcile loop is cancelled
+		case <-ctx.Done():
+			r.Log.Info("context cancelled, stopping scraping metrics")
+			return
+
+		case <-time.After(r.ScrapeInterval):
+			r.enqueueGenericEventsFromMetrics(ctx)
+		}
+	}
+}
+
+func (r *StorageAutoscalerScraper) enqueueGenericEventsFromMetrics(ctx context.Context) {
+	// query to get the osd usage
+	osdUsageQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))"
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdUsageQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return
+	}
+	// update the sync map with the highest value of osd usage per device class
+	updateSyncMap(metrics, r.SyncMap)
+	// get the storage autoscaler objects
+	objectList, err := getStorageAutoScalerObjects(ctx, r.Client, r.Log, r.SyncMap)
+	if err != nil {
+		r.Log.Error(err, "failed to get resource objects")
+		return
+	}
+	// filter the objects for scaling
+	filteredObjectList := filterObjectsForScaling(objectList, r.SyncMap, r.Log)
+
+	// send generic events for the filtered objects
+	sendGenericEvent(filteredObjectList, r.EventCh, r.Log)
+}
+
+// update the sync map with the highest value of osd usage per device class
+func updateSyncMap(metrics model.Vector, syncMap *sync.Map) {
+	// create a map of device class to highest osd usage
+	deviceClassUsage := make(map[string]float64)
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		if _, ok := deviceClassUsage[string(deviceClass)]; ok {
+			if float64(osd.Value) > deviceClassUsage[string(deviceClass)] {
+				deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+			}
+		} else {
+			deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+		}
+	}
+
+	// update the sync map
+	for deviceClass, usage := range deviceClassUsage {
+		syncMap.Store(deviceClass, usage)
+	}
+}
+
+func getStorageAutoScalerObjects(ctx context.Context, client client.Client, log logr.Logger, syncMap *sync.Map) (*ocsv1.StorageAutoScalerList, error) {
```

Comment:
Consider defining these methods with the receiver. Other methods can also consider the same change, to simplify all the parameters. 
```suggestion
func (r *StorageAutoscalerScraper) getStorageAutoScalerObjects() (*ocsv1.StorageAutoScalerList, error) {
```

Reply:
Actually, I moved from that to this parameterized function as per above comments, should we go back to that?

Reply:
Are you referring to [this comment](https://github.com/red-hat-storage/ocs-operator/pull/3071/files#r1993912277) regarding refactoring for unit testing or which comment? Not sure why we wouldn't want to use the receiver to reduce the complexity of parameters.

Reply:
I don't see `log` or `syncMap` used in the function, so maybe those can be removed to reduce the params

Reply:
Mostly all these comments https://github.com/red-hat-storage/ocs-operator/pull/3071#discussion_r1990812166 https://github.com/red-hat-storage/ocs-operator/pull/3071#discussion_r1990739044, etc from which I refactor it back.